### PR TITLE
[MDEP-628] Parameter `fileMappers` for unpack path rewriting

### PR DESCRIPTION
**Pull request by mkarg**
Solving [MDEP-628](https://issues.apache.org/jira/browse/MDEP-628).

The new parameter `fileMappers` (default: `null`) can be set to an implementation of the `org.codehaus.plexus.components.io.filemappers.FileMapper` interface to rewrite the target path of each unpacked file.

This is useful in case prefixes of target files names within the target directory shall be added (using `PrefixFileMapper`), changed or omitted (using `RegExpFileMapper`).

### CODE REVIEW

**Comment by plamentotev** [comment]
What is the use case that this feature covers? I just wonder if you need complex and powerful way to organize files, isn't assembly plugin (with `dir` archive) more suitable?

**Comment by mkarg** [comment]
Believe me, it was the perfectly correct choice to provide *exactly this* solution, and we really need this complete freedom of mapping! ;-)

First of all, the assembly plugin has the same problem: It cannot *freely* rewrite target paths, but this is what we actually solved. Second, as we do not want to create assemblies at all, it would simply be a misuse; what we actually want is to unpack dependencies, so the correct way *is* to dependency:unpack and nothing else.

In cour case, what we needed is to invoke external tools as part of our build process which expects a particular directory hierarchy. The source of the files in that hierarchy come from several *external* dependencies (ZIP files), i. e. their content is not under our control. The problem is: The tools expect a completely different hierarchy than the one found in the ZIPs. For example, each ZIP has some kinde of parent folders structure inside, but the tool wants a different parent structure:
```
Dependency ZIP A:
/a/b/C

Dependency ZIP B:
/d/e/F

Tool can only process *exactly this* folder structure:
/u/v/w/C
/x/y/z/F
```
The solution of extending the dependency plugin is clean and straightforward: We now can simply add the RegExMapper config within each `<ArtifactItem>` in the unpack config.

So to sum up: It was the perfectly correct way to fork and extend the dependecy plugin. Everything else would just be a bad hack. ;-)

**Comment by plamentotev** [comment]
Yes, you're right that in such case it does make sense to use the dependency plugin instead of the assembly plugin.

**Comment by mkarg** [comment]
@plamentotev Review please. :-)

**Comment by plamentotev** [comment]
@mkarg I don't have commit rights and I'm not very familiar with the
plugin code base so it would be better if somebody else reviews it.
I've asked about this pull request because recently there was similar
contribution to Plexus Archiver about file mappers  and I was curious
it it is the same use case (but it turned out to be a different one).
Sorry for bringing a bit of noise in your pull request :)

**Comment by mkarg** [comment]
Could someone else review please? :-)

**Comment by plamentotev** [comment]
@mkarg Plexus Archiver 3.7.0 is now released so you can update the dependency.

**Comment by mkarg** [comment]
@khmarbaise @hboutemy I would be happy if you could review this PR and consider for merging. :-)

**Comment by rfscholte** [comment]
A ping like this helps. I've had a look at the patch, but I am missing a test to confirm this feature. Probably easiest solved by providing an Integration Test under `src/it/projects`, executed with `mvn verify -Prun-its`

**Comment by mkarg** [comment]
@rfscholte Amended existing integration test to verify mapping works well, see d47a4abbe0817cef27ea078ec7b35ca4e81b451b.

**Comment by rfscholte** [comment]
Could you copy the unpack-dependencies instead of adjusting it? We need to ensure that the default 1:1 extraction keeps working as expected.

**Comment by mkarg** [comment]
@rfscholte Done. :-)

**Comment by mkarg** [comment]
@rfscholte Thanks! :-) :tada:
